In [ ]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

# Atividade de vendas no Brasil

## Entregável 1 NumPy

### A. Criando arrays

In [ ]:
#Lendo o arquivo csv com pandas
df = pd.read_csv("vendas_brasil_1.csv")

#Criando arrays para as colunas numéricas do dataset usando numpy
preco_unitario_np = df['preco_unitario'].to_numpy()

custo_unitario_np = df['custo_unitario'].to_numpy()

quantidade_np = df['quantidade'].to_numpy()

desconto_pct_np = df['desconto_pct'].to_numpy()

receita_total_np = df['receita_total'].to_numpy()

frete_np = df['frete'].to_numpy()
print("Dataset")
print(df)
print(f'Os arrays criados a partir deste dataset foram:\nPreço unitario:{preco_unitario_np}\ncusto unitario:{custo_unitario_np}\ndesconto pct:{desconto_pct_np}\nquantidade:{quantidade_np}\nreceita total:{receita_total_np}\nfrete:{frete_np} ')


Dataset
       data_venda  mes          cidade  canal_venda    categoria  \
0      2022-04-13    4  Rio de Janeiro   App Mobile  Eletrônicos   
1      2023-03-12    3          Manaus     Site Web    Vestuário   
2      2022-09-28    9         Goiânia   App Mobile    Vestuário   
3      2022-04-17    4       Fortaleza     Site Web    Vestuário   
4      2022-03-13    3  Belo Horizonte   App Mobile     Esportes   
...           ...  ...             ...          ...          ...   
37612  2023-12-31    3    Porto Alegre   App Mobile  Alimentação   
37613  2023-12-31    9        Salvador     Site Web  Alimentação   
37614  2023-12-31    2        Salvador  Loja Física  Alimentação   
37615  2023-12-31   11       São Paulo     Site Web     Esportes   
37616  2023-12-31    4       Fortaleza   Televendas  Eletrônicos   

               produto       vendedor fidelidade_cliente  preco_unitario  \
0             Notebook    Bruno Souza           Diamante          739.39   
1             Camiseta 

### B. Cálculos com operações vetorizadas

In [ ]:
#calculando valores importantes sobre o dataset
valor_bruto = preco_unitario_np * quantidade_np
valor_desconto = valor_bruto * (desconto_pct_np/100)
custo_total = custo_unitario_np * quantidade_np
lucro_estimado = receita_total_np - custo_total

print(f'valor bruto dos produtos: {valor_bruto}')
print(f'desconto dos produtos: {valor_desconto}')
print(f'custo total da produção dos produtos: {custo_total}')
print(f'lucro estimado da venda de cada produto: {lucro_estimado}')


valor bruto dos produtos: [7393.9   937.4  1065.54 ...  386.05 1379.63 4229.64]
desconto dos produtos: [384.4828    5.6244    7.45878 ...   0.      150.37967 422.964  ]
custo total da produção dos produtos: [5617.5   675.75  859.6  ...  258.58 1170.61 3482.76]
lucro estimado da venda de cada produto: [1315.98  254.53  198.9  ...  125.86  789.57  312.97]


### C. Máscaras booleanas

In [ ]:
#Filtrando situações inusitadas do dataset

media_receita = np.mean(receita_total_np)
vendas_acima_media = receita_total_np[receita_total_np > media_receita]

vendas_prejuizo = lucro_estimado[lucro_estimado < 0]

media_desconto = np.mean(desconto_pct_np)
vendas_desconto_alta = desconto_pct_np[desconto_pct_np > media_desconto]

print(f'quantidade de vendas com receita acima da média: {len(vendas_acima_media)}')
print(f'quantidade vendas que geraram prejuízo: {len(vendas_prejuizo)}')
print(f'quantidade de vendas com desconto acima da média: {len(vendas_desconto_alta)}')
print(f'\nquantidade de vendas totais: {len(df['data_venda'])}')

quantidade de vendas com receita acima da média: 9236
quantidade vendas que geraram prejuízo: 2168
quantidade de vendas com desconto acima da média: 17853

quantidade de vendas totais: 37617


### D. Agregações

In [ ]:
#calculando medidas gerais para ánalise

media_das_receitas = np.mean(receita_total_np)

maior_receita = np.max(receita_total_np)

menor_receita = np.min(receita_total_np)

media_dos_lucros = np.mean(lucro_estimado)

quantidade_vendas_prejuizo = len(vendas_prejuizo)

print(f'media das receitas: {media_das_receitas}')
print(f'maior receita: {maior_receita}')
print(f'menor receita: {menor_receita}')
print(f'media dos lucros: {media_dos_lucros}')
print(f'quantidade de vendas com prejuizo: {quantidade_vendas_prejuizo}')

media das receitas: 1891.6881729005502
maior receita: 58430.28
menor receita: 3.7
media dos lucros: 513.187152883005
quantidade de vendas com prejuizo: 2168


### Conclusões gerais


Observando o dataset apenas com funções matematicas e de array do numpy podemos observar algumas inconsistências.

Dentro das vendas que análisamos vemos que temos alguns outliers tendo em vista que a média das receitas é 1891 enquanto a maior e menor receita que vimos são respectivamente 58430 e 3.

Podemos notar tambem que existem 2168 produtos onde o lucro obtido é negativo, gerando necessidade de uma maior atenção e ánalise em tais objetos

## Entregável 2 Pandas


### A. Primeiro olhar no dataset

In [ ]:
#explorando de forma geral o dataset usando pandas

print("Exibindo as 5 primeiras linhas do dataset:\n")
df.head()

Exibindo as 5 primeiras linhas do dataset:



,data_venda,mes,cidade,canal_venda,categoria,produto,vendedor,fidelidade_cliente,preco_unitario,custo_unitario,...,quantidade,desconto_pct,receita_total,frete,avaliacao_cliente,prazo_entrega_dias,devolucao,entrega_no_prazo,alta_temporada,flag_outlier_preco
0,2022-04-13,4,Rio de Janeiro,App Mobile,Eletrônicos,Notebook,Bruno Souza,Diamante,739.39,561.75,...,10,5.2,6933.48,8.68,4.92,5.0,False,True,False,False
1,2023-03-12,3,Manaus,Site Web,Vestuário,Camiseta,Elisa Costa,Bronze,187.48,135.15,...,5,0.6,930.28,NaN,4.98,19.0,True,False,False,False
2,2022-09-28,9,Goiânia,App Mobile,Vestuário,Tênis,João Vieira,Bronze,76.11,61.40,...,14,0.7,1058.50,7.98,3.47,NaN,True,True,False,False
3,2022-04-17,4,Fortaleza,Site Web,Vestuário,Calça Jeans,Carla Mendes,Bronze,145.86,88.53,...,11,0.6,1592.43,20.49,4.50,16.0,False,True,False,False
4,2022-03-13,3,Belo Horizonte,App Mobile,Esportes,Bicicleta,João Vieira,Bronze,133.92,107.25,...,1,19.5,107.64,22.89,3.49,13.0,False,True,False,False


In [ ]:
print("Exibindo a estrutura do nosso dataset")
df.info()

Exibindo a estrutura do nosso dataset
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 37617 entries, 0 to 37616
Data columns (total 21 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   data_venda          37617 non-null  object 
 1   mes                 37617 non-null  int64  
 2   cidade              37617 non-null  object 
 3   canal_venda         37567 non-null  object 
 4   categoria           37617 non-null  object 
 5   produto             37617 non-null  object 
 6   vendedor            37577 non-null  object 
 7   fidelidade_cliente  37617 non-null  object 
 8   preco_unitario      37617 non-null  float64
 9   custo_unitario      37617 non-null  float64
 10  margem_pct          37535 non-null  float64
 11  quantidade          37617 non-null  int64  
 12  desconto_pct        37617 non-null  float64
 13  receita_total       37617 non-null  float64
 14  frete               35738 non-null  float64
 15  avaliacao_clien

In [ ]:
print("Exibindo um resumo das informações estatisticas do nosso dataset")
df.describe()

Exibindo um resumo das informações estatisticas do nosso dataset


,mes,preco_unitario,custo_unitario,margem_pct,quantidade,desconto_pct,receita_total,frete,avaliacao_cliente,prazo_entrega_dias
count,37617.000000,37617.000000,37617.000000,37535.000000,37617.000000,37617.000000,37617.000000,35738.000000,34662.000000,36113.000000
mean,6.517691,253.214933,183.620999,22.714776,7.481750,9.494048,1891.688173,29.885737,3.912094,10.017113
std,3.439296,316.742367,231.376757,8.670213,4.322273,9.097484,2972.445759,24.765161,0.932256,5.477487
min,1.000000,4.940000,3.010000,-0.570000,1.000000,0.000000,3.700000,3.750000,1.000000,1.000000
25%,4.000000,68.960000,49.550000,16.380000,4.000000,0.300000,364.280000,12.190000,3.490000,5.000000
50%,7.000000,111.300000,80.990000,22.700000,7.000000,9.000000,789.570000,22.250000,4.010000,10.000000
75%,9.000000,257.680000,186.570000,29.250000,11.000000,19.000000,1853.210000,39.480000,4.520000,15.000000
max,12.000000,9008.600000,6578.260000,44.270000,161.000000,26.000000,58430.280000,243.820000,5.000000,19.000000


In [ ]:
print("exibindo o número de linhas e colunas do nosso dataset")
df.shape

exibindo o número de linhas e colunas do nosso dataset


(37617, 21)

In [ ]:
print("exibindo o número de nulos do nosso dataset por cada coluna")
df.isna().sum()

exibindo o número de nulos do nosso dataset por cada coluna


,0
data_venda,0
mes,0
cidade,0
canal_venda,50
categoria,0
produto,0
vendedor,40
fidelidade_cliente,0
preco_unitario,0
custo_unitario,0


In [ ]:
print("exibi o total de dados nulos no dataset")
df.isna().sum().sum()

exibi o total de dados nulos no dataset


np.int64(6510)

1- Existem 37617 linhas e 21 colunas

2- Sim, existe um total de 6510 dados nulos espalhados pelas colunas do dataset

3- as colunas:

*   mes
*   preco_unitario
*   custo_unitario
*   margem_pct
*   quantidade
*   desconto_pct
*   receita_total
*   frete
*   avaliação_cliente
*   prazo_entrega_dia

4- as colunas:


*   data_venda
*   cidade
*   canal_venda
*   categoria
*   produto
*   vendedor
*   fidelidade_cliente
*   devolucao
*   entrega_no_prazo
*   alta_temporada
*   flag_outlier_preco





### B. Limpeza e tratamento

In [ ]:
#convertendo a coluna data para datatype
df['data_venda'] = pd.to_datetime(df['data_venda'], format='mixed')

#padronizando os nomes das cidades
df['cidade'] = df['cidade'].str.upper()
mapeamento_cidades = {
    'RIO': 'RIO DE JANEIRO',
    'RJ': 'RIO DE JANEIRO',
    'RIO DEJANEIRO': 'RIO DE JANEIRO',
    'CURTIBA': 'CURITIBA',
    'SP': 'SÃO PAULO',
    'S. PAULO': 'SÃO PAULO',
    'SAO PAULO': 'SÃO PAULO',
    'BH': 'BELO HORIZONTE',
    'B. HORIZONTE': 'BELO HORIZONTE',
    'SLAVADOR': 'SALVADOR'
    }
df['cidade'] = df['cidade'].str.strip()
df['cidade'] = df['cidade'].replace(mapeamento_cidades)
df['cidade'].value_counts()

#Analisando a coluna avaliacao_cliente para decidir a melhor forma de substituir dados faltantes
"""
print(f'Analisando a coluna "avaliacao_cliente"')
print(f'média: {df['avaliacao_cliente'].mean()}')
print(f'maior coluna: {df['avaliacao_cliente'].max()}')
print(f'menor coluna: {df['avaliacao_cliente'].min()}')
"""
media_avaliacao_clientes = df['avaliacao_cliente'].mean()
df['avaliacao_cliente'] = df['avaliacao_cliente'].fillna(media_avaliacao_clientes)
df.isna().sum()

#Analisando a coluna frete para decidir a melhor forma de substituir dados faltantes
'''
print(f'Analisando a coluna "frete"')
print(f'{df['frete']}')
'''
df['frete'] = df['frete'].fillna(0)
df.isna().sum()
df['data_venda']

,data_venda
0,2022-04-13
1,2023-03-12
2,2022-09-28
3,2022-04-17
4,2022-03-13
...,...
37612,2023-12-31
37613,2023-12-31
37614,2023-12-31
37615,2023-12-31


Decidi substituir os dados faltantes da coluna 'frete' por 0 pois acho plausível supor que os dados nulos que aparecem são devidos a não necessidade de entrega do produto fazendo o frete não existir.

Como a coluna 'avaliacao_clientes' não possuia outliers grandes a melhor opção para substituir os dados faltantes é muda-los pela média da própria coluna.

Dessa maneira evitamos um viés muito grande na ánalise do noss0 dataset

### C. Criação de novas colunas

In [ ]:
#Criando uma coluna para o ano das vendas
df['ano_venda'] = df['data_venda'].dt.year
df[['ano_venda','data_venda']]

#Criando uma coluna para o lucro total
df['lucro_total'] = lucro_estimado #Variavel "lucro_estimado" criada em 1.B Calculos com operações vetorizadas
df['lucro_total']

#Criando uma coluna para receita liquida
df['receita_liquida'] = df['receita_total'] - df['desconto_pct'] - df['frete']
df[['receita_liquida', 'receita_total']].head()

#Criando uma coluna para faixa de receita
df['faixa_receita'] = df['receita_total'].apply(lambda x: 'alta' if x > media_das_receitas else 'baixa')
df[['faixa_receita', 'receita_total']]

,faixa_receita,receita_total
0,alta,6933.48
1,baixa,930.28
2,baixa,1058.50
3,baixa,1592.43
4,baixa,107.64
...,...,...
37612,baixa,524.49
37613,baixa,820.52
37614,baixa,384.44
37615,alta,1960.18


### D. Filtros

In [ ]:
#Criando um filtro para saber quais receitas foram altas
receita_alta = df[df['faixa_receita'] == 'alta']
receita_alta

#Filtro para saber quais vendas tiveram devolução
vendas_com_devolucao = df[df['devolucao'] == True]
df['devolucao'].value_counts() #Nenhum dado fora do comum
vendas_com_devolucao

#Filtrando quais vendas foram da categoria Alimentação
vendas_alimentacao = df[df['categoria'] == 'Alimentação']
vendas_alimentacao

print(f'Os filtros criados foram:\nReceitas altas:\n{receita_alta}\nitens devolvidos\n{vendas_com_devolucao}\nquais vendas foram para Alimentação:\n{vendas_alimentacao}')

Os filtros criados foram:
Receitas altas:
      data_venda  mes          cidade  canal_venda    categoria  \
0     2022-04-13    4  RIO DE JANEIRO   App Mobile  Eletrônicos   
11    2022-11-27   11       SÃO PAULO     Site Web     Esportes   
15    2022-04-10    4        SALVADOR  Marketplace  Eletrônicos   
17    2022-05-11    5  BELO HORIZONTE   App Mobile  Eletrônicos   
20    2022-12-10   12          RECIFE     Site Web  Eletrônicos   
...          ...  ...             ...          ...          ...   
37593 2023-12-31    2          MANAUS     Site Web  Eletrônicos   
37603 2023-12-31    9  RIO DE JANEIRO   App Mobile  Eletrônicos   
37605 2023-12-31    3       FORTALEZA  Marketplace     Esportes   
37615 2023-12-31   11       SÃO PAULO     Site Web     Esportes   
37616 2023-12-31    4       FORTALEZA   Televendas  Eletrônicos   

              produto      vendedor fidelidade_cliente  preco_unitario  \
0            Notebook   Bruno Souza           Diamante          739.39   
11   

Com os filtros pudemos encontrar quais foram as receitas que tiveram um alto valor(valor acima da média), temos uma noção de quais vendas tiveram devolução e quantas e quais vendas foram da categoria de Alimetnação

### E. Agrupamentos com groupby

In [ ]:
#Calculando a receita que cada categoria gerou
receita_categoria = df.groupby('categoria')['receita_total'].sum()
receita_categoria

#Descobrindo qual canal de venda vendeu em maior quantidade
canal_mais_vendas = df.groupby('canal_venda')['quantidade'].sum()
canal_mais_vendas

#Analisando qual cidade teve maior receita
cidade_maior_receita = df.groupby('cidade')['receita_total'].sum()
cidade_maior_receita

,receita_total
cidade,
BELO HORIZONTE,6914747.61
CURITIBA,6993175.07
FORTALEZA,7242081.82
GOIÂNIA,7178113.69
MANAUS,7459161.23
PORTO ALEGRE,7048428.17
RECIFE,7662241.79
RIO DE JANEIRO,6880176.34
SALVADOR,6918656.01


### F. Merge

In [ ]:
#Criando uma tabela de regiões para ligar cada cidade do dataset a uma região
regioes = {
    'cidade': ['RIO DE JANEIRO', 'SÃO PAULO', 'BELO HORIZONTE', 'CURITIBA', 'FORTALEZA', 'GOIÂNIA', 'MANAUS', 'PORTO ALEGRE','RECIFE', 'SALVADOR'],
    'regioes': ['SUDESTE', 'SUDESTE', 'SUDESTE', 'SUL', 'NORDESTE', 'CENTRO-OESTE', 'NORTE', 'SUL', 'NORDESTE', 'NORDESTE']
}
df_regioes = pd.DataFrame(regioes)
df_regioes

df_atualizado = pd.merge(df, df_regioes, on='cidade', how='left')
receita_por_regiao = df_atualizado.groupby('regioes')['receita_total'].sum()
receita_por_regiao

,receita_total
regioes,
CENTRO-OESTE,7178113.69
NORDESTE,21822979.62
NORTE,7459161.23
SUDESTE,20657776.22
SUL,14041603.24


## Entregável 3 Seaborn